### PR TITLE
feat: pure esm

### PR DESCRIPTION
**Pull request by jxom**
This PR converts `wagmi` from bundling both CJS + ESM, to bundling pure ESM. 

I have published the packages to NPM as a snapshot release under the tag `esm-only` (`npm i wagmi@esm-only`).

This will probably affect tooling that is sensitive to importing ESM such as Jest and Remix. We might need to create some additional guides on how to use wagmi with these tools as we move to pure ESM.

### CODE REVIEW

File: examples/remix/remix.config.js [hunk 1]
```diff
@@ -8,4 +8,5 @@ module.exports = {
   serverBuildDirectory: 'build',
   devServerPort: 8002,
   ignoredRouteFiles: ['.*'],
+  serverDependenciesToBundle: [/^@?wagmi.*/, 'zustand/vanilla'],
```
Comment (jxom) [inline examples/remix/remix.config.js:11]: https://remix.run/docs/en/v1/pages/gotchas#importing-esm-packages

File: patches/@preconstruct__cli@2.2.1.patch [hunk 2]
```diff
@@ -0,0 +1,949 @@
+# Source: https://github.com/preconstruct/preconstruct/compare/main...jxom:preconstruct:tmm/esm-support
```
Comment (jxom) [inline patches/@preconstruct__cli@2.2.1.patch:1]: Source for the patch.

File: packages/core/chains/package.json [hunk 3]
```diff
@@ -1,4 +1,4 @@
 {
-  "main": "dist/wagmi-core-chains.cjs.js",
-  "module": "dist/wagmi-core-chains.esm.js"
+  "main": "dist/wagmi-core-chains.esm.js",
```
Comment (jxom) [inline packages/core/chains/package.json:2]: Ideally, we don't even need these `package.json`'s, but we may need them for a couple of reasons:

- Some tooling (Jest) don't respect the `exports` field, and
- Setting `moduleResolution` to anything other than `node16` or `nodenext` in `tsconfig.json` breaks type resolution. However, if we do set `moduleResolution` to one of those values, we need to append the `.js` extension to every relative import (yuck). There are long running issues for this [here](https://github.com/microsoft/TypeScript/issues/50152) and [here](https://github.com/microsoft/TypeScript/issues/49083).

File: packages/core/src/actions/accounts/fetchBalance.ts [hunk 4]
```diff
@@ -1,5 +1,5 @@
-import { BigNumber, logger } from 'ethers/lib/ethers'
-import { Logger, formatUnits, isAddress } from 'ethers/lib/utils'
+import { BigNumber, logger } from 'ethers/lib/ethers.js'
```
Comment (jxom) [inline packages/core/src/actions/accounts/fetchBalance.ts:1]: Is there a reason why we aren't using `ethers` here? Should I move to that?

**Comment by tmm** [comment]
opening up PR with different approach shortly

**Pull request closed without merging**